docs/source/development/defining_new_models.md

```diff
@@ -0,0 +1,221 @@
+# How to add a new model to the Virtual Rainforest
+
+The Virtual Rainforest is designed to be modular, meaning that the set of models to be
+used in a particular run is configurable at the start of the simulation. We are starting
+out by defining a core set of models (`abiotic`, `animals`, `plants` and `soil`), which
+will generally all be included for the vast majority of simulations. In future, it might
+be desirable to include models for other aspects of rainforests (e.g. `freshwater`), or
+to include multiple modelling approaches for a process. When this happens a new model
+should be created. This page will set out the process for adding a new model to the
+Virtual Rainforest in a manner that allows it to be used appropriately by the `core`
+simulation functionality.
+
+## Define a new `Model` class
+
+You should first start by defining a new folder for your model (within
+`virtual_rainforest/models/`).
+
+```bash
+mkdir virtual_rainforest/models/freshwater
+```
+
+Within this folder a `python` script defining the model should be created.
+
+```bash
+touch virtual_rainforest/models/freshwater/model.py
+```
+
+This script must import the [`BaseModel` class](../api/core/model.md) as new class must
+inherit from this abstract base class.
+
+```python
+from virtual_rainforest.core.model import BaseModel
+```
+
+The new model class is created using a class method, this means that a model name must
+be provided when class inheritance occurs, so that the model can automatically be added
+to the registry under that name. This name does not strictly have to match the
+`self.name` attribute, but it generally should.
```

Comment:
Maybe the registry should extract `cls.name` rather than having duplicate locations - there is no reason why these shouldn't be identical?

Reply:
Yeah relatively easy fix. I've added the change to the pull request

Reply:
Honestly, I'd do it the other way around. So rather than synchronising two locations:

```python
    @classmethod
    def __init_subclass__(cls):
        """Method that adds new model classes to the model registry."""

        if getattr(cls, 'name') is None:
             LOGGER.critical("Model is missing the model name class attribute")
        if not isinstance(cls.name, str)
             LOGGER.critical("Model name is not a string")

        # Add the grid type to the registry
        if cls.name in MODEL_REGISTRY:
            LOGGER.warning(
                "Model with name %s already exists and is being replaced", cls.name 
            )
        MODEL_REGISTRY[cls.name ] = cls
```

Reply:
I think there's a way of decorating a class attribute so that is part of the ABC and hence cannot be left empty. Still might have bogus content though.

Reply:
And maybe use `f"The 'name' class attribute is not set in model class {cls.__name__}"`

Reply:
I kind of prefer supplying a model name though, think it makes it clearer that a something other than normal class inheritance is being used.

Reply:
I'm only keeping the `self.name` in `BaseModel` so that the base model unit tests don't break, could just remove that name entirely, and rewrite the unit tests a bit

Reply:
The problem though, is that they should be identical, and I think it is _much_ stranger to have a class attribute written in the code that is then replaced with another value.

Reply:
I think `cls.name` is important. When the model is created and running, that is the only place where the model name is stored and hence can be used as `self.name` in logger calls etc. And that name is _fundamentally_ an attribute of the class.

Reply:
Also it isn't great to have a class attribute that doesn't appear in the code with a docstring. So, it should be there in the subclass definitions.

Reply:
I think I might be misunderstanding something here. Are `cls.name` and `self.name` different things?

Reply:
I was imagining that `cls.name` was `self.name` for the class in question

Reply:
They do refer to the same thing, but in the context of the `class` method, which uses `cls` as an argument to indicate that it is referring to the properties of the class itself, rather than `self.name` referring to the properties of an _instance_ of the class. The point of class attributes specifically is that property value is set across all possible instances of the class.

Reply:
But in defining a subclass, which has a class attribute `name` (maybe we should using `model_name`), I think that attribute should be set and docstring in the model code, rather than set via an argument to `__init_subclass`. It won't appear in the class attributes for example. Given that needs to happen, it makes sense to use it to register the class?

Reply:
So the issue with the current way I'm doing it is that only class instances produced using `__init_subclass__` would end with a name populated?

Reply:
If someone called `SoilModel(**kwargs)`, it would create an instance without a name?

Reply:
Or at the moment I think it would be called "base", but was planning on deleting that

Reply:
The use of `class ModelName(BaseModel)` _guarantees_ that `BaseModel.__init_subclass__` will be called, so `cls.name` would be populated on any BaseModel subclass. **But** that is an obscure way to populate a class attribute and those are typically included and documented as:

```python
class ModelName(BaseModel):

     model_name = "ModelName"
     """The model name for use in registering the model and logging"
``

Reply:
No - the name is _only_ set in the creation of the class itself. So your `soil.model` object _creates_ the `SoilModel` object, and at that point the subclass itself (not a class instance) is registered using `__init_subclass__`.

If someone creates an _instance_ using `SoilModel(**kwargs)`, then that creates an instance, with the `name` class attribute populated when the package was loaded. The difference here is really where it is clearer to set that attribute, and I think the direct option above is better for clarity and having attributes appear in rendered docstrings.

Reply:
Happy to switch it around if you think that makes sense! My only hesitation is that it feels like a fairly big functional change to the code (which would require new tests, etc). Wonder if that is sensible to be adding to a pull request about docs, or whether I should make a new issue + PR for it?

Reply:
@dalonsoa @alexdewar - Thoughts? It's not a large change but does need code change in the BaseModel and tests. Seems like a small PR to me. And also - am I talking nonsense?

Reply:
@davidorme Sounds sensible to me.

I would do it more simply though (I didn't know `__init_subclass__` was a thing till now). I'd just have a check in `BaseModel`'s `__init__` along the lines of:

```py
if not hasattr(self, "name"):
    raise ValueError("Models must have a name attribute")
````